File: refactor/backend/src/app/api/routes/strategy.py
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from app.api.deps import get_strategy_service
from app.services.strategy_service import StrategyService

router = APIRouter(prefix="/strategy")


class CognitionDistillRequest(BaseModel):
    session_id: str = Field(min_length=1)
    start_index: int | None = Field(default=None, ge=1)
    end_index: int | None = Field(default=None, ge=1)
    title: str | None = None


class CognitionReviewRequest(BaseModel):
    action: str = Field(pattern="^(approve|reject|edit)$")
    reviewer: str = Field(min_length=1)
    editor_notes: str | None = None
    edited_markdown: str | None = None


class StrategyExtractRequest(BaseModel):
    strategy_type: str = Field(pattern="^(analysis|trading)$")
    source_scope: str = "indexed_memos"
    prompt_ref: str | None = None


class StrategyPublishRequest(BaseModel):
    backtest_job_id: str | None = None
    proposal_id: str | None = None


class StrategyRollbackRequest(BaseModel):
    reason: str | None = None


class StrategyBindRequest(BaseModel):
    flow_id: str = Field(min_length=1)
    prompt_refs: list[str] = Field(default_factory=list)
    prompt_lock_mode: str | None = Field(default=None, pattern="^(strict|lenient)$")
    effective_scope: dict[str, Any] = Field(default_factory=dict)


@router.post("/cognition/distill", status_code=201)
def distill_cognition(
    request: CognitionDistillRequest,
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    try:
        return service.distill_cognition(
            session_id=request.session_id,
            start_index=request.start_index,
            end_index=request.end_index,
            title=request.title,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/cognition/{memo_id}/review")
def review_cognition(
    memo_id: str,
    request: CognitionReviewRequest,
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    try:
        return service.review_cognition(
            memo_id=memo_id,
            action=request.action,
            reviewer=request.reviewer,
            editor_notes=request.editor_notes,
            edited_markdown=request.edited_markdown,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/extract", status_code=201)
def extract_strategy(
    request: StrategyExtractRequest,
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    try:
        return service.extract_strategy(
            strategy_type=request.strategy_type,
            source_scope=request.source_scope,
            prompt_ref=request.prompt_ref,
        )
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/versions")
def list_strategy_versions(
    strategy_type: str | None = Query(default=None),
    status: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    return service.list_versions(
        strategy_type=strategy_type,
        status=status,
        limit=limit,
    )


@router.post("/{strategy_id}/publish")
def publish_strategy(
    strategy_id: str,
    request: StrategyPublishRequest,
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    try:
        return service.publish_strategy(
            strategy_id=strategy_id,
            backtest_job_id=request.backtest_job_id,
            proposal_id=request.proposal_id,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/{strategy_id}/bind", status_code=201)
def bind_strategy(
    strategy_id: str,
    request: StrategyBindRequest,
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    try:
        return service.bind_strategy(
            strategy_id=strategy_id,
            flow_id=request.flow_id,
            prompt_refs=request.prompt_refs,
            prompt_lock_mode=request.prompt_lock_mode,
            effective_scope=request.effective_scope,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/bindings")
def list_strategy_bindings(
    flow_id: str | None = Query(default=None),
    strategy_id: str | None = Query(default=None),
    status: str | None = Query(default=None),
    limit: int = Query(default=100, ge=1, le=500),
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    return service.list_bindings(
        flow_id=flow_id,
        strategy_id=strategy_id,
        status=status,
        limit=limit,
    )


@router.post("/{strategy_id}/rollback")
def rollback_strategy(
    strategy_id: str,
    request: StrategyRollbackRequest,
    service: StrategyService = Depends(get_strategy_service),
) -> dict[str, Any]:
    try:
        return service.rollback_strategy(
            strategy_id=strategy_id,
            reason=request.reason,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc

File: refactor/frontend/src/lib/services/strategy.ts
import { getJson, postJson } from "../api";
import type {
  CognitionMemo,
  StrategyArtifact,
  StrategyBinding,
  StrategyBindingsResponse,
  StrategyVersionsResponse,
} from "../types";

export type DistillCognitionInput = {
  session_id: string;
  start_index?: number;
  end_index?: number;
  title?: string;
};

export type ReviewCognitionInput = {
  action: "approve" | "reject" | "edit";
  reviewer: string;
  editor_notes?: string;
  edited_markdown?: string;
};

export type ExtractStrategyInput = {
  strategy_type: "analysis" | "trading";
  source_scope?: string;
  prompt_ref?: string;
};

export type PublishStrategyInput = {
  backtest_job_id?: string;
  proposal_id?: string;
};

export type BindStrategyInput = {
  flow_id: string;
  prompt_refs?: string[];
  prompt_lock_mode?: "strict" | "lenient";
  effective_scope?: Record<string, unknown>;
};

export type RollbackStrategyInput = {
  reason?: string;
};

export type ListVersionsInput = {
  strategy_type?: string;
  status?: string;
  limit?: number;
};

export type ListBindingsInput = {
  flow_id?: string;
  strategy_id?: string;
  status?: string;
  limit?: number;
};

function buildQuery(input: Record<string, string | number | undefined>): string {
  const params = new URLSearchParams();
  Object.entries(input).forEach(([key, value]) => {
    if (value === undefined || value === "") {
      return;
    }
    params.set(key, String(value));
  });
  return params.toString();
}

export async function distillCognition(input: DistillCognitionInput): Promise<CognitionMemo> {
  return postJson<CognitionMemo>("/strategy/cognition/distill", input);
}

export async function reviewCognition(memoId: string, input: ReviewCognitionInput): Promise<CognitionMemo> {
  return postJson<CognitionMemo>(`/strategy/cognition/${memoId}/review`, input);
}

export async function extractStrategy(input: ExtractStrategyInput): Promise<StrategyArtifact> {
  return postJson<StrategyArtifact>("/strategy/extract", input);
}

export async function listStrategyVersions(input: ListVersionsInput = {}): Promise<StrategyVersionsResponse> {
  const query = buildQuery({
    strategy_type: input.strategy_type,
    status: input.status,
    limit: input.limit,
  });
  const suffix = query ? `?${query}` : "";
  return getJson<StrategyVersionsResponse>(`/strategy/versions${suffix}`);
}

export async function publishStrategy(strategyId: string, input: PublishStrategyInput): Promise<StrategyArtifact> {
  return postJson<StrategyArtifact>(`/strategy/${strategyId}/publish`, input);
}

export async function bindStrategy(strategyId: string, input: BindStrategyInput): Promise<StrategyBinding> {
  return postJson<StrategyBinding>(`/strategy/${strategyId}/bind`, input);
}

export async function listStrategyBindings(input: ListBindingsInput = {}): Promise<StrategyBindingsResponse> {
  const query = buildQuery({
    flow_id: input.flow_id,
    strategy_id: input.strategy_id,
    status: input.status,
    limit: input.limit,
  });
  const suffix = query ? `?${query}` : "";
  return getJson<StrategyBindingsResponse>(`/strategy/bindings${suffix}`);
}

export async function rollbackStrategy(strategyId: string, input: RollbackStrategyInput): Promise<StrategyArtifact> {
  return postJson<StrategyArtifact>(`/strategy/${strategyId}/rollback`, input);
}
